jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
You're right about the index and making sure we're not reading beyong the array, sorry I forgot about that... Another reason why we went with the iterator. You don't have to think about that, because the `jerryx_arg_iterator_pop/peak()` will worry about this for you :P

> I'd like to pick your brain what's envisioned.

In the context of this [challenging use case], I was thinking of a "transform" that mutates the next `jerry_arg_t`. In that specific example, to update the destination pointer for the function transform, to point to `&native_obj->callback` (only after having run the first transformer, `native_obj` is pointing to something valid). I'm still thinking about whether that's this is a good solution or not though... (any suggestions welcome ;))
Yes, you need access to the structure of the iterator to do that, but it would be a tiny *addition* compared to having to change the transformer function signature. Yes, you also need to overrule `const` in that particular transformer (not nice I know).

@HBehrens: Do you have strong arguments for the iterator approach? You suggested it to me and @jiangzidong in the first place.

[challenging use case]: https://github.com/jerryscript-project/jerryscript/pull/1740#issuecomment-297300966